llpc/tool/llpcComputePipelineBuilder.cpp

```diff
@@ -0,0 +1,159 @@
+Error ComputePipelineBuilder::build() {
+  CompileInfo &compileInfo = getCompileInfo();
+
+  Expected<BinaryData> pipelineOrErr = buildComputePipeline();
+  if (Error err = pipelineOrErr.takeError())
+    return err;
+
+  Result result = decodePipelineBinary(&*pipelineOrErr, &compileInfo);
+  if (result != Result::Success)
+    return createResultError(result, "Failed to decode pipeline");
+
+  return Error::success();
+}
+
+// =====================================================================================================================
+// Build the compute pipeline.
+//
+// @returns : Pipeline binary data on success, `llvm::ResultError` on failure.
+Expected<BinaryData> ComputePipelineBuilder::buildComputePipeline() {
+  CompileInfo &compileInfo = getCompileInfo();
+  assert(compileInfo.shaderModuleDatas.size() == 1);
+
+  StandaloneCompiler::ShaderModuleData &moduleData = compileInfo.shaderModuleDatas[0];
+  assert(moduleData.shaderStage == ShaderStageCompute);
+
+  ComputePipelineBuildInfo *pipelineInfo = &compileInfo.compPipelineInfo;
+  ComputePipelineBuildOut *pipelineOut = &compileInfo.compPipelineOut;
+
+  PipelineShaderInfo *shaderInfo = &pipelineInfo->cs;
+  const ShaderModuleBuildOut *shaderOut = &moduleData.shaderOut;
+
+  // If entry target is not specified, use the one from command line option.
+  if (!shaderInfo->pEntryTarget)
+    shaderInfo->pEntryTarget = moduleData.entryPoint.c_str();
+
+  shaderInfo->entryStage = ShaderStageCompute;
+  shaderInfo->pModuleData = shaderOut->pModuleData;
+
+  // If not compiling from pipeline, lay out user data now.
+  if (compileInfo.doAutoLayout) {
+    ResourceMappingNodeMap nodeSets;
+    unsigned pushConstSize = 0;
+    doAutoLayoutDesc(ShaderStageCompute, moduleData.spirvBin, nullptr, shaderInfo, nodeSets, pushConstSize,
+                     /*autoLayoutDesc =*/compileInfo.autoLayoutDesc);
+
+    buildTopLevelMapping(ShaderStageComputeBit, nodeSets, pushConstSize, &pipelineInfo->resourceMapping,
+                         compileInfo.autoLayoutDesc);
+  }
+
+  pipelineInfo->pInstance = nullptr; // Placeholder, unused.
+  pipelineInfo->pUserData = &compileInfo.pipelineBuf;
+  pipelineInfo->pfnOutputAlloc = allocateBuffer;
+  pipelineInfo->unlinked = compileInfo.unlinked;
+  pipelineInfo->options.robustBufferAccess = compileInfo.robustBufferAccess;
+  pipelineInfo->options.enableRelocatableShaderElf = compileInfo.relocatableShaderElf;
+  pipelineInfo->options.scalarBlockLayout = compileInfo.scalarBlockLayout;
+  pipelineInfo->options.enableScratchAccessBoundsChecks = compileInfo.scratchAccessBoundsChecks;
+
+  PipelineBuildInfo localPipelineInfo = {};
+  localPipelineInfo.pComputeInfo = pipelineInfo;
+  void *pipelineDumpHandle = runPreBuildActions(localPipelineInfo);
+  auto onExit = make_scope_exit([&] { runPostBuildActions(pipelineDumpHandle, {pipelineOut->pipelineBin}); });
+
+  ICompiler &compiler = getCompiler();
```

Comment:
Don't know if this is part of this PR (if not, just ignore it), but the reference to the compiler is not necessary here.

Reply:
I don't understand, `compiler` is used in the next line

Reply:
It's a matter of taste, but I'd prefer to use `getCompiler()` directly if there's only a single use of the variable.

Reply:
Ah, makes sense